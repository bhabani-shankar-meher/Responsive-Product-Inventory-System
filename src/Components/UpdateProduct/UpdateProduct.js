import { useEffect, useState } from "react";
import { useParams, useNavigate } from "react-router-dom";
import { Formik, Form, Field } from "formik";
import { Button, CircularProgress } from "@mui/material";
import Box from "@mui/material/Box";
import Grid from "@mui/material/Grid";
import { TextField } from "formik-mui";
import * as Yup from "yup";

const UpdateProduct = () => {
  const { id } = useParams();
  let navigate = useNavigate();
  const [selectedItem, setSelectedItem] = useState("");
  useEffect(() => {
    fetch(`http://localhost:4000/products/${id}`).then((response) =>
      response.json().then((data) => setSelectedItem(data))
    );
  }, [id]);

  const productSchema = Yup.object().shape({
    productName: Yup.string()
      .min(2, "Too Short!")
      .required("Product name is Required"),
    quantity: Yup.number().required("Quantity is Required"),
    price: Yup.string().required("Price is Required"),
  });

  return (
    <Formik
      enableReinitialize={true}
      initialValues={{
        productName: selectedItem && selectedItem.productName,
        quantity: selectedItem && selectedItem.quantity,
        price: selectedItem && selectedItem.price,
      }}
      validationSchema={productSchema}
      onSubmit={(values, { setSubmitting }) => {
        try {
          if (selectedItem) {
            setSubmitting(false);
            values.updatedDate = new Date().toLocaleString();
            fetch(`http://localhost:4000/products/${id}`, {
              method: "PATCH",
              headers: {
                "Content-Type": "application/json",
              },
              body: JSON.stringify(values, null, 2),
            });
            navigate("/", { replace: true });
          }
        } catch (exception) {
          console.log(exception);
        }
      }}
    >
      {({ submitForm, isSubmitting }) => (
        <Form>
          <Box sx={{ width: "70%" }} mt={5} mb={3} style={{ marginLeft: "15%" }}>
            <Grid
              container
              rowSpacing={3}
              columnSpacing={{ xs: 1, sm: 2, md: 3 }}
            >
              <Grid item xs={12} md={12}>
                <Field
                  component={TextField}
                  name="productName"
                  type="text"
                  label="Product Name"
                  style={{ width: "100%" }}
                />
              </Grid>
              <Grid item xs={12} md={12}>
                <Field
                  component={TextField}
                  name="quantity"
                  type="number"
                  label="Quantity"
                  style={{ width: "100%" }}
                />
              </Grid>
              <Grid item xs={12} md={12}>
                <Field
                  component={TextField}
                  name="price"
                  type="text"
                  label="Price (INR)"
                  style={{ width: "100%" }}
                />
              </Grid>
            </Grid>
          </Box>
          <Grid item xs={12} md={12}>
            {isSubmitting && <CircularProgress />}
          </Grid>
          <Box sx={{ width: "70%" }} style={{ marginLeft: "15%" }} >
            <Grid item xs={12} md={12}>
              <Button
                variant="contained"
                color="primary"
                disabled={isSubmitting}
                onClick={submitForm}
              >
                Submit
              </Button>
              <Button
                variant="outlined"
                color="secondary"
                style={{ marginLeft: "5px" }}
                onClick={() => navigate("/", { replace: true })}
              >
                Back
              </Button>
            </Grid>
          </Box>
        </Form>
      )}
    </Formik>
  );
};

export default UpdateProduct;
