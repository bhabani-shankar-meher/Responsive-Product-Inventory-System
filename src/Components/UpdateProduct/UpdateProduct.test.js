import { render, screen } from "@testing-library/react";
import { create } from "react-test-renderer";
import UpdateProduct from "./UpdateProduct";
jest.mock("react-router-dom", () => ({
    ...jest.requireActual("react-router-dom"),
    useNavigate: () => jest.fn(),
}));
describe("UPDATEPRODUCT PAGE SNAPSHOT", () => {
    test("UpdateProduct Page Snapshot Executed Successfully", () => {
        const updateProductSnapshot = create(<UpdateProduct />);
        expect(updateProductSnapshot).toMatchSnapshot();
    });
});

describe("UPDATEPRODUCT PAGE TEST", () => {
    test("Product Name Label", () => {
        render(<UpdateProduct />);
        const label = screen.getAllByText(/product name/i);
        expect(label).not.toHaveLength(0);
    });
    test("Quantity Label", () => {
        render(<UpdateProduct />);
        const label = screen.getAllByText(/quantity/i);
        expect(label).not.toHaveLength(0);
    });
    test("Price Label", () => {
        render(<UpdateProduct />);
        const label = screen.getAllByText(/price/i);
        expect(label).not.toHaveLength(0);
    });
    test("Submit Button Test", () => {
        render(<UpdateProduct />);
        expect(
            screen.getByRole("button", {
                name: /submit/i,
            })
        ).toBeInTheDocument();
    });
    test("Back Button Test", () => {
        render(<UpdateProduct />);
        expect(
            screen.getByRole("button", {
                name: /back/i,
            })
        ).toBeInTheDocument();
    });
});
