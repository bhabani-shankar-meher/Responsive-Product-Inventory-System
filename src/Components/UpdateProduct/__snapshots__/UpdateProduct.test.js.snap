// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`UPDATEPRODUCT PAGE SNAPSHOT UpdateProduct Page Snapshot Executed Successfully 1`] = `
<form
  action="#"
  onReset={[Function]}
  onSubmit={[Function]}
>
  <div
    className="MuiBox-root css-1ahqaq8"
    style={
      Object {
        "marginLeft": "15%",
      }
    }
  >
    <div
      className="MuiGrid-root MuiGrid-container css-qwfiov-MuiGrid-root"
    >
      <div
        className="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-12 MuiGrid-grid-md-12 css-1idn90j-MuiGrid-root"
      >
        <div
          className="MuiFormControl-root MuiTextField-root css-1u3bzj6-MuiFormControl-root-MuiTextField-root"
          style={
            Object {
              "width": "100%",
            }
          }
        >
          <label
            className="MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-outlined MuiFormLabel-root MuiFormLabel-colorPrimary css-14s5rfu-MuiFormLabel-root-MuiInputLabel-root"
            data-shrink={false}
            htmlFor=":r0:"
            id=":r0:-label"
          >
            Product Name
          </label>
          <div
            className="MuiOutlinedInput-root MuiInputBase-root MuiInputBase-colorPrimary MuiInputBase-formControl css-9ddj71-MuiInputBase-root-MuiOutlinedInput-root"
            onClick={[Function]}
          >
            <input
              aria-invalid={false}
              autoFocus={false}
              className="MuiOutlinedInput-input MuiInputBase-input css-1t8l2tu-MuiInputBase-input-MuiOutlinedInput-input"
              disabled={false}
              id=":r0:"
              name="productName"
              onAnimationStart={[Function]}
              onBlur={[Function]}
              onChange={[Function]}
              onFocus={[Function]}
              required={false}
              type="text"
              value=""
            />
            <fieldset
              aria-hidden={true}
              className="MuiOutlinedInput-notchedOutline css-1d3z3hw-MuiOutlinedInput-notchedOutline"
            >
              <legend
                className="css-1ftyaf0"
              >
                <span>
                  Product Name
                </span>
              </legend>
            </fieldset>
          </div>
        </div>
      </div>
      <div
        className="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-12 MuiGrid-grid-md-12 css-1idn90j-MuiGrid-root"
      >
        <div
          className="MuiFormControl-root MuiTextField-root css-1u3bzj6-MuiFormControl-root-MuiTextField-root"
          style={
            Object {
              "width": "100%",
            }
          }
        >
          <label
            className="MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-outlined MuiFormLabel-root MuiFormLabel-colorPrimary css-14s5rfu-MuiFormLabel-root-MuiInputLabel-root"
            data-shrink={false}
            htmlFor=":r1:"
            id=":r1:-label"
          >
            Quantity
          </label>
          <div
            className="MuiOutlinedInput-root MuiInputBase-root MuiInputBase-colorPrimary MuiInputBase-formControl css-9ddj71-MuiInputBase-root-MuiOutlinedInput-root"
            onClick={[Function]}
          >
            <input
              aria-invalid={false}
              autoFocus={false}
              className="MuiOutlinedInput-input MuiInputBase-input css-1t8l2tu-MuiInputBase-input-MuiOutlinedInput-input"
              disabled={false}
              id=":r1:"
              name="quantity"
              onAnimationStart={[Function]}
              onBlur={[Function]}
              onChange={[Function]}
              onFocus={[Function]}
              required={false}
              type="number"
              value=""
            />
            <fieldset
              aria-hidden={true}
              className="MuiOutlinedInput-notchedOutline css-1d3z3hw-MuiOutlinedInput-notchedOutline"
            >
              <legend
                className="css-1ftyaf0"
              >
                <span>
                  Quantity
                </span>
              </legend>
            </fieldset>
          </div>
        </div>
      </div>
      <div
        className="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-12 MuiGrid-grid-md-12 css-1idn90j-MuiGrid-root"
      >
        <div
          className="MuiFormControl-root MuiTextField-root css-1u3bzj6-MuiFormControl-root-MuiTextField-root"
          style={
            Object {
              "width": "100%",
            }
          }
        >
          <label
            className="MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-outlined MuiFormLabel-root MuiFormLabel-colorPrimary css-14s5rfu-MuiFormLabel-root-MuiInputLabel-root"
            data-shrink={false}
            htmlFor=":r2:"
            id=":r2:-label"
          >
            Price (INR)
          </label>
          <div
            className="MuiOutlinedInput-root MuiInputBase-root MuiInputBase-colorPrimary MuiInputBase-formControl css-9ddj71-MuiInputBase-root-MuiOutlinedInput-root"
            onClick={[Function]}
          >
            <input
              aria-invalid={false}
              autoFocus={false}
              className="MuiOutlinedInput-input MuiInputBase-input css-1t8l2tu-MuiInputBase-input-MuiOutlinedInput-input"
              disabled={false}
              id=":r2:"
              name="price"
              onAnimationStart={[Function]}
              onBlur={[Function]}
              onChange={[Function]}
              onFocus={[Function]}
              required={false}
              type="text"
              value=""
            />
            <fieldset
              aria-hidden={true}
              className="MuiOutlinedInput-notchedOutline css-1d3z3hw-MuiOutlinedInput-notchedOutline"
            >
              <legend
                className="css-1ftyaf0"
              >
                <span>
                  Price (INR)
                </span>
              </legend>
            </fieldset>
          </div>
        </div>
      </div>
    </div>
  </div>
  <div
    className="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-12 MuiGrid-grid-md-12 css-13i4rnv-MuiGrid-root"
  />
  <div
    className="MuiBox-root css-6gs9hh"
    style={
      Object {
        "marginLeft": "15%",
      }
    }
  >
    <div
      className="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-12 MuiGrid-grid-md-12 css-13i4rnv-MuiGrid-root"
    >
      <button
        className="MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-sizeMedium MuiButton-containedSizeMedium MuiButtonBase-root  css-sghohy-MuiButtonBase-root-MuiButton-root"
        disabled={false}
        onBlur={[Function]}
        onClick={[Function]}
        onContextMenu={[Function]}
        onDragLeave={[Function]}
        onFocus={[Function]}
        onKeyDown={[Function]}
        onKeyUp={[Function]}
        onMouseDown={[Function]}
        onMouseLeave={[Function]}
        onMouseUp={[Function]}
        onTouchEnd={[Function]}
        onTouchMove={[Function]}
        onTouchStart={[Function]}
        tabIndex={0}
        type="button"
      >
        Submit
      </button>
      <button
        className="MuiButton-root MuiButton-outlined MuiButton-outlinedSecondary MuiButton-sizeMedium MuiButton-outlinedSizeMedium MuiButtonBase-root  css-wnpvqa-MuiButtonBase-root-MuiButton-root"
        disabled={false}
        onBlur={[Function]}
        onClick={[Function]}
        onContextMenu={[Function]}
        onDragLeave={[Function]}
        onFocus={[Function]}
        onKeyDown={[Function]}
        onKeyUp={[Function]}
        onMouseDown={[Function]}
        onMouseLeave={[Function]}
        onMouseUp={[Function]}
        onTouchEnd={[Function]}
        onTouchMove={[Function]}
        onTouchStart={[Function]}
        style={
          Object {
            "marginLeft": "5px",
          }
        }
        tabIndex={0}
        type="button"
      >
        Back
      </button>
    </div>
  </div>
</form>
`;
